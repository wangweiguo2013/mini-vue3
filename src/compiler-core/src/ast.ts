export const enum NodeTypes {
    INTERPOLATION, // 差值
    SIMPLE_EXPRESSION
}
