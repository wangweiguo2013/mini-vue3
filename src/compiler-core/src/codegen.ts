export const codegen = (root) => {
    const context = createCodeGenContext()
    const { push } = context

    const codegenNode = root.codegenNode
    const signature = ['ctx', '_cache']

    const functionName = 'render'
    push(` function ${functionName}(${signature.join(', ')}){`)

    //  render函数参数
    genNode(codegenNode, context)
    push('}')
    return { code: context.code }
}

function genNode(codegenNode, context) {
    const { push } = context
    push(`return '${codegenNode.content}'`)
}

function createCodeGenContext() {
    const context = {
        code: '',
        push(source) {
            context.code += source
        }
    }
    return context
}
