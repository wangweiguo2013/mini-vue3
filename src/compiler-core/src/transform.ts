/**
 * 通过传入plugin的方式，在外部控制需要操作的函数
 * 降低程序耦合
 * 测试的时候也方便按需测试
 */

export const transform = (root, options = {}) => {
    const context = createTransformContext(root, options)
    traverseNode(root, context)

    createCodegenNode(root)
    return root
}

function createCodegenNode(root: any) {
    root.codegenNode = root.children[0]
}

function traverseNode(node, context) {
    const nodeTransforms = context.nodeTransforms
    nodeTransforms.forEach((transform) => {
        transform(node)
    })

    traverseChildren(node, context)
}

function traverseChildren(node, context) {
    if (node.children) {
        for (let i = 0; i < node.children.length; i++) {
            const child = node.children[i]
            console.log('node ----', node)
            traverseNode(child, context)
        }
    }
}

function createTransformContext(root: any, options) {
    return {
        root,
        nodeTransforms: options.nodeTransforms || []
    }
}
