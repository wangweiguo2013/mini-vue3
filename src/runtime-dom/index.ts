import { createRenderer } from '../runtime-core'

export const createElement = (type) => {
    return document.createElement(type)
}
const isEvent = (key: string) => /^on[A-Z]/.test(key)

export const patchProp = (el, key, preValue, nextValue) => {
    console.log('update prop, new value is %s, old value is %s', nextValue, preValue)
    if (isEvent(key)) {
        const eventName = key.slice(2).toLocaleLowerCase()
        el.addEventListener(eventName, nextValue)
    } else if (nextValue === null || nextValue === undefined) {
        el.removeAttribute(key)
    } else {
        el.setAttribute(key, nextValue)
    }
}

export const insert = (el, parent) => {
    parent.append(el)
}

export const setElementText = (container, text) => {
    container.textContent = text
}
export const remove = (el) => {
    const parent = el.parentNode
    parent & parent.remove(el)
}

const renderer: any = createRenderer({
    createElement,
    patchProp,
    insert,
    setElementText,
    remove
})

export const createApp = (...args) => {
    return renderer.createApp(...args)
}

export * from '../runtime-core'
