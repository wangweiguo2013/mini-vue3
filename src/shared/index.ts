export const extend = Object.assign

export const isObject = (raw) => {
    return raw !== undefined && raw !== null && typeof raw === 'object'
}

export const hasChanged = (value, newValue) => {
    return !Object.is(value, newValue)
}

export const capitalize = (str: string) => {
    return str ? str.charAt(0).toUpperCase() + str.slice(1) : ''
}

export const camelize = (str: string) => {
    return str.replace(/-\w?/g, (_, c: string) => {
        return _.slice(1).toUpperCase()
    })
}
