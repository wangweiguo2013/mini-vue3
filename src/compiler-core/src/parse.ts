import { NodeTypes } from './ast'

export const baseParse = (content: string) => {
    const context = createParserContext(content)
    return createRoot(parseChildren(context))
}
function createParserContext(content: string) {
    return {
        source: content
    }
}

function createRoot(children) {
    return {
        children
    }
}

function parseChildren(context): any {
    const nodes: any[] = []

    let node
    if (context.source.startsWith('{{')) {
        node = parseInterpolation(context)
    }

    nodes.push(node)

    return nodes
}
function parseInterpolation(context) {
    const openDelimiter = '{{'
    const closeDelimiter = '}}'

    const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)

    advancedBy(context, openDelimiter.length)

    const rawContentLength = closeIndex - openDelimiter.length

    const rawContent = context.source.slice(0, rawContentLength)
    const content = rawContent.trim()

    advancedBy(context, rawContentLength + closeDelimiter.length)

    return {
        type: NodeTypes.INTERPOLATION,
        content: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: content
        }
    }
}

function advancedBy(context, length) {
    context.source = context.source.slice(length)
}
