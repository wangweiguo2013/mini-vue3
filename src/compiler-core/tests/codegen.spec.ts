import { codegen } from '../src/codegen'
import { baseParse } from '../src/parse'
import { transform } from '../src/transform'

describe('code generate', () => {
    it('happy path', () => {
        const ast = baseParse('hi')
        transform(ast)
        const { code } = codegen(ast)
        expect(code).toMatchSnapshot()
    })
})
