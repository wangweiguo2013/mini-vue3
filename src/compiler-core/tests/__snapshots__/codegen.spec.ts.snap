// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code generate happy path 1`] = `" function render(ctx, _cache){return 'hi'}"`;
